egraph.rkt

```diff
@@ -94,24 +135,22 @@
 ;; an existing node or otherwise attached to some node to be
 ;; completely added to the egraph.
 (define (mk-enode! eg expr)
-  (if (hash-has-key? (egraph-expr->parent eg) expr)
-      (let ([res (hash-ref (egraph-expr->parent eg) expr)])
-	(pack-leader res))
-      (let* ([expr* (if (not (list? expr)) expr
-			(cons (car expr)
-			      (map pack-leader (cdr expr))))]
-	     [en (new-enode expr* (egraph-cnt eg))]
-	     [leader->iexprs (egraph-leader->iexprs eg)])
- 	(set-egraph-cnt! eg (add1 (egraph-cnt eg)))
-	(hash-set! leader->iexprs en (mutable-set))
-	(when (list? expr*)
-	  (for ([suben (in-list (cdr expr*))])
-	    (set-add! (hash-ref leader->iexprs (pack-leader suben))
-		      expr*)))
-	(hash-set! (egraph-expr->parent eg)
-		   expr*
-		   en)
-	en)))
+  (let ([expr* (update-en-expr expr)])
```

Comment:
**Bug 1**
Here we update the expression first thing, before checking if it is already in the egraph.